common/event_cache.go

```diff
@@ -0,0 +1,202 @@
+package common
+
+import (
+	"fmt"
+	"github.com/openziti/ziti/common/pb/edge_ctrl_pb"
+	"sync"
+)
+
+type OnStoreSuccess func(index uint64, event *edge_ctrl_pb.DataState_Event)
+
+type EventCache interface {
+	// Store allows storage of an event and execution of an onSuccess callback while the event cache remains locked.
+	// onSuccess may be nil. This function is blocking.
+	Store(event *edge_ctrl_pb.DataState_Event, onSuccess OnStoreSuccess) error
+
+	// CurrentIndex returns the latest event index applied. This function is blocking.
+	CurrentIndex() (uint64, bool)
+
+	// ReplayFrom returns an array of events from startIndex and true if the replay may be facilitated.
+	// An empty slice and true is returned in cases where the requested startIndex is the current index.
+	// An empty slice and false is returned in cases where the replay cannot be facilitated.
+	// This function is blocking.
+	ReplayFrom(startIndex uint64) ([]*edge_ctrl_pb.DataState_Event, bool)
+
+	// WhileLocked allows the execution of arbitrary functionality while the event cache is locked. This function
+	// is blocking.
+	WhileLocked(func())
+}
+
+// ForgetfulEventCache does not store events or support replaying. It tracks
+// the event index and that is it. It is a stand in for LoggingEventCache
+// when replaying events is not expected (i.e. in routers)
+type ForgetfulEventCache struct {
+	lock  sync.Mutex
+	index *uint64
+}
+
+func NewForgetfulEventCache() *ForgetfulEventCache {
+	return &ForgetfulEventCache{}
+}
+
+func (cache *ForgetfulEventCache) WhileLocked(callback func()) {
+	cache.lock.Lock()
+	defer cache.lock.Unlock()
+
+	callback()
+}
+
+func (cache *ForgetfulEventCache) Store(event *edge_ctrl_pb.DataState_Event, onSuccess OnStoreSuccess) error {
+	cache.lock.Lock()
+	defer cache.lock.Unlock()
+
```

Comment:
Should this allow stale events to be ignored and only error if we see the index jump? Are stale events likely to be a regular occurrence?

Reply:
Yes. Taken care of.